### ISSUE TITLE
Implement a dispatched model for compilation of the CPU capabilities

### ISSUE THREAD

**Issue by K2** [RoaringBitmap/CRoaring#113]
Since we have a range of optimizations currently available and likely to add more/improve over time, the complexity of building with the appropriate configuration can be tricky.  

Ideally, we would have a dispatching library that is able to select our optimal implementation at run time based on the CPU feature set.  Most libraries I’ve seen that offer dispatched models use environment variables to disable or adjust things on the fly.

For GCC we could use https://gcc.gnu.org/wiki/FunctionMultiVersioning, MSVC is a bit more complicated and it may be inevitable that there will be some function pointers to glue things together, I’ve started with some intrinsic routines to get going also and will be evaluating some options in keeping with pure C (no templates).  It'd be nice to have a simple pattern that worked for all compiler/build targets.

**Comment by lemire** [comment]
I think that it is desirable and quite interesting. How to make it work portably is, of course, the main difficulty.

**Comment by lemire** [comment]
Google Pik has SIMD-based runtime dispatch: 

https://github.com/google/pik/blob/master/simd/dispatch.cc

**Comment by K2** [comment]
I took a look and it seems they have a pretty similar impression that most who go this route use either some intrinsic glue or C++ templates.  About 80% of the dispatched code I see is setup with templates.  Is some very light C++ use OK for us or should we always consider pure C?  I was even looking at the possibility of using some JIT from clang to glue up the appropriate callout's ;)...

The majority of pik's design doc seems to be here, https://github.com/google/pik/tree/master/simd, down in the prior designs section they talk about some of their reasoning and reduced CPU support envelope.

I adapted the Linux test cases to Windows and was evaluating them under some of the Intel profiles (seems they have a give away for open source projects if you look also..)

This run may of had profiling hooks slowing it down a tad, but I think I recall you mentioning another set of test cases that may have more coverage?  I was not sure if these were them or not, I had to expand some of them to generate more data as some of tap's didnt register w/o some big-data-esq levels of execution I guess ;)

C:\Users\files\Source\git\RoaringBit\build\benchmarks\Release>real_bitmaps_benchmark.exe C:\Users\files\Source\git\RoaringBit\benchmarks\realdata\
Loaded 1800 bitmaps from directory C:\Users\files\Source\git\RoaringBit\benchmarks\realdata\
Creating 1800 bitmaps took 748652575 cycles
Copying and freeing 1800 bitmaps took 6867095 cycles
 1799 successive bitmaps intersections took 39300501 cycles
 1799 successive bitmaps unions took 46095795 cycles
 1799 successive in-place bitmaps intersections took 44307588 cycles
 1799 successive in-place bitmaps unions took 58031522 cycles
Iterating over 1800 bitmaps and 44247613 values took 623591460 cycles

**Comment by K2** [comment]
Oh and really quick I don't know if I mentioned this one or not but it jogged my memory, extracting some of the micro-op data from https://github.com/google/EXEgesis should give us/anybody definitive guidelines on how to maximize the various pipeline stages, memory bandwidth and any out-of-order assembly.  I generated the data a while back but they seem pretty active so keeping an eye on the info you can get from their generators should help improve any asm code over time.

**Comment by lemire** [comment]
For our immediate needs, we just want to dynamically disable/enable vectorization when the runtime-detected processor does not meet the requirements. This does not require a lot of new code. In fact, it could be done by touching fewer than 100 lines, I think.

One can havemore ambitious goals, of course...

**Comment by lemire** [comment]
For an extensive benchmark, please see...

https://github.com/RoaringBitmap/CBitmapCompetition

**Comment by SalvatorePreviti** [comment]
If needed, I use a CPU ID detection in the NodeJS port of the project.
I compile the whole loadable module 3 times (plain, sse4.2, avx) and then chose the module to load depending on the cpu_id result. Not feasible for all environment but the cpu_id code should work on every platform.

https://github.com/SalvatorePreviti/roaring-node/blob/master/src/cpuinfo/cpuinfo.cpp

**Comment by lemire** [comment]
That's clever and useful.

**Comment by SalvatorePreviti** [comment]
Just an idea... Can we use the typecode used in the containers to redirect to the right plain/see/avx2 implementation? We would not need to do pointer tables or add another indirection level

**Comment by K2** [comment]
Cool JS @SalvatorePreviti.  BTW along the lines of run-time / selection of machine code implementation... Would it make any seance to consider use cases around side channel attack prevention?  I noticed some interesting patterns for this https://arxiv.org/pdf/1806.05179.pdf and a pre-meltdown era paper https://eprint.iacr.org/2017/699.pdf (though it suggests polymorphic code as a viable defence)... not sure on how/if the protection of a bitmap's state is a consideration to users of Roaring... just figured I'd float the idea.

**Comment by SalvatorePreviti** [comment]
@K2 Not sure it would be an issue, the use of AVX instructions per se is not an issue. And I trust your implementation :) 

In the NodeJS version I added a detection for the ROARING_DISABLE_AVX environment variable, if somebody is concerned about the use of AVX or SSE4.2 is easy to disable, this leaves to the user the choice. I am personally using it in production with AVX support enabled, not too concerned.

Would be nice also for CRoaring to provide a way to disable also at runtime the use of AVX and SSE4.2

**Comment by andreigudkov** [comment]
Can someone please explain the idea behind ROARING_DISABLE_AVX flag? It looks redundant to me. There are already builtin ways of dealing with instruction sets. Both GCC and VC++ define macroes `__AVX__`/`__AVX2__` if target architecture supports them:
````
#ifdef __AVX2__
  void bitset_union() { ... }
#else
  void bitset_union() { ... }
#endif
````

To select architecture, GCC provides `-march`/`-mtune` options, and VC++ provides `/arch:AVX` and `/arch:AVX2`. For example, if I wanted to use roaring in a cluster with skylake CPUs, I would squeeze out max performance with:
```
cmake -DCMAKE_C_FLAGS='-march=skylake -mtune=skylake' .. && make
```

**Comment by SalvatorePreviti** [comment]
ROARING_DISABLE_AVX (and ROARING_DISABLE_SSE42) in the nodejs version is an environment variable that is evaluated at runtime,  it alters the cpu_id detection mechanism disabling AVX. __AVX2__ or DISABLEAVX defines are evaluated only at compile time

**Comment by lemire** [comment]
See related discussion in another project https://github.com/lemire/simdjson/issues/90

cc @TkTech

**Comment by TkTech** [comment]
It looks like @SalvatorePreviti is doing the same cpuid check in his NodeJS port, but it is missing extended control register check to see if YMM registers are enabled. This will cause failures in the case where the processor supports AVX2 but the OS hasn't enabled it.

**Comment by lemire** [comment]
This is related to the issue of being Python wheels... 
https://github.com/Ezibenroc/PyRoaringBitMap/issues/45

**Comment by SalvatorePreviti** [comment]
@TkTech thanks, I will have a look on this!

**Comment by SalvatorePreviti** [comment]
Found this, https://hg.mozilla.org/mozilla-central/file/64bab5cbb9b6/mozglue/build/SSE.cpp#l190 - will try to use the same strategy on the NodeJS cpu detection

**Comment by andreigudkov** [comment]
I was thinking about implementing dispatch by compiling roaring multiple times with different optimization flags and creating vtable for all public functions. This is in contrast to another popular approach when dispatch happens only for a handful of internal functions which have multiple implementations (e.g. `bitset_container_and()`). Below example demonstrates proposed changes and also introduces error codes and logging function.

This is how users are supposed to use new API:
    
    // Create roaring environment; typically, this is done only once per thread.
    // First argument is custom logging function, second one -- copy on write flag
    roaring_bitmap_env_t* env = roaring_bitmap_create_env(NULL, false);
    
    // all roaring functions take pointer to roaring_bitmap_env* as their first argument
    roaring_bitmap_t* r = roaring_bitmap_create(env);
    if (env->error) {
      exit(1);
    }
    roaring_bitmap_add(env, r, 5);
    if (env->error) {
      exit(1);
    }
    roaring_bitmap_free(env, r);

    roaring_bitmap_free_env(env);

API viible through headers:

    typedef enum {
      OK = 0,
      NO_MEMORY,
      BUFFER_UNDERFLOW, // serialization
      BAD_COOKIE, // serialization
      ...
    } roaring_error_t;
    
    // Type for user-provided logging function
    typedef void (*roaring_log_fn)(const char* msg);
    
    typedef struct {
      roaring_bitmap_t* (roaring_bitmap_create_fn*) (roaring_env_t*);
      void              (roaring_bitmap_free_fn*  ) (roaring_env_t*, roaring_bitmap_t*);
      void              (roaring_bitmap_add_fn*   ) (roaring_env_t*, roaring_bitmap_t*, uint32_t);
      // ... long list of API functions ...
    } roaring_vtable_t;
    
    typedef struct {
      const roaring_bitmap_vtable_t* vtable;
    
      // Users may override this function; by default it is NULL
      roaring_log_fn log_fn;
    
      // allow newly created bitmaps to share some data with bitmaps they are created from
      bool copy_on_write;
    
      // error code of the last roaring operation
      roaring_error_t error;
    } roaring_bitmap_env_t;

Hidden inside sources:

    void roaring_bitmap_add__avx2(roaring_bitmap_env* env, roaring_bitmap* rb, uint32_t value) {
      // ...
      // example how to use logger and error codes
      if (malloc(...) == NULL) {
        if (env->log_fn) {
          env->log_fn("Couldn't allocate %zu bytes at", sizeof(...));
        }
        env->error = NOT_ENOUGH_MEMORY;
        return;
      }
      // ...
      env->error = OK;
    }
    void roaring_bitmap_add__x64(roaring_bitmap_env* env, roaring_bitmap* rb, uint32_t value) {
      // ...
    }
    
    // dispatch function
    inline void roaring_bitmap_add(roaring_bitmap_env* env, roaring_bitmap_t* rb, uint32_t value) {
      env->vtable->roaring_bitmap_add(env, rb, value);
    }
    
    static const roaring_vtable_t avx2_vtable = {
      .roaring_bitmap_create_fn = ...,
      .roaring_bitmap_free_fn = ...,
      .roaring_bitmap_add_fn = roaring_bitmap_add__avx2,
      ...
    };
    
    static const roaring_vtable_t x64_vtable = {
      ...
    };
    
    roaring_bitmap_env_t* roaring_bitmap_create_env(logger_fn* logger, bool copy_on_write) {
      roaring_bitmap_env_t *env = malloc(...);
      env->vtable = cpu_supports_avx2() ? avx2_vtable : x64_vtable;
      env->logger = logger;
      env->copy_on_write = copy_on_write;
      env->error = OK;
      return env;
    }

There is list of problems that must be solved:

* Think how to make cmake compile multiple versions of the same code with different optimization settings and with different function suffixes, ideally without a lot of bloat. (Something like `gcc -DFUNCTION_SUFFIX="__avx2" -mavx2 source.c`)
* Think how to adapt `roaring_bitmap_env_t` to C++ wrapper.
* Think how to force roaring to use suboptimal implementation for performance testing (environment variable? `ROARING_IMPL=x64noavx ./roaring_benchmark`)

**Comment by SalvatorePreviti** [comment]
@TkTech - do you mind have a look at this, to check wether I am forgetting some kind of tests? https://github.com/SalvatorePreviti/roaring-node/blob/be23ca0c70752d295fc43d1e2a1f5d3060d96a18/src/cpuinfo/cpuinfo.cpp

**Comment by lemire** [comment]
@andreigudkov 

Maybe you are right, but this feels intrusive to me.

My impression is that other software out there has dynamic dispatch without putting as large a burden on the programmer (e.g., https://github.com/google/pik). cc @jan-wassenberg 


Here is my own view of how to do this (it might be hopelessly naive).

Have a runtime constant that tells you whether you have a recent x64 compiler (RUNTIME_AVX2). We only use AVX2-acceleration in a few instances. Only in these instances... do...

1. Check whether RUNTIME_AVX2 has been initialized, if not initialize it. That's a single predictable branch. The initialization of the constant should not matter.
2. Then branch on RUNTIME_AVX2 but only, evidently, if the code was compiled with support for AVX2. Again, that's a single predictable branch.

This requires you to compile the code only once. It is entirely transparent to the library user (they don't even need to know that this is happening). 

Also, the way this works, you can disable fancy optimizations *even* if the processor supports them and you have compiled with the fancy optimizations... just change the value RUNTIME_AVX2 at runtime.

Note: RUNTIME_AVX2 is not a good name.

Maybe I am being stupid, but that's how I'd do it.

**Comment by andreigudkov** [comment]
Just a small remark.

We don't know whether target system has AVX2 or no. That's why in your approach the whole library must be compiled without AVX2 support, so that compiler doesn't autovectorize code with AVX2. And only a small subset of functions with AVX2 intrinsics (which should be moved into separate files) must be compiled with AVX2.

**Comment by lemire** [comment]
> We don't know whether target system has AVX2 or no.

Yes, we don't. That's why we should check it at runtime. But this requires only a single variable. This variable can initialized with a CPU check once. There is no reason that the library user should need to know about this variable.

What am I missing?

I don't think we disagree @andreigudkov on the facts, I am only asking why the solution should involve changing the API (=making it harder to use).

**Comment by TkTech** [comment]
@salvatorepreviti happy to take a look after work!

@lemire @andriegudkov what you're discussing is called function multi-versioning in gcc and is available in gcc, ICC, and clang.

**Comment by lemire** [comment]
@TkTech Yes, we could solve this with function attributes *if* we were happy to set aside Visual Studio and other compilers... But we are not happy with that so we need to do it ourselves.

**Comment by lemire** [comment]
@TkTech  See the top of the issue where @K2 wrote...

>  For GCC we could use https://gcc.gnu.org/wiki/FunctionMultiVersioning, MSVC is a bit more complicated and it may be inevitable that there will be some function pointers to glue things together, I’ve started with some intrinsic routines to get going also and will be evaluating some options in keeping with pure C (no templates). It'd be nice to have a simple pattern that worked for all compiler/build targets.


What we would really need is some *standard* approach but there is none.

**Comment by andreigudkov** [comment]
Access to AVX2-intrinsic functions is guarded by the global variable. But what is about other functions? Since we do not know the capabilities of the target platform, we can only compile them with basic x64 features. I am worried that this will be suboptimal.

(Obviously, I do not insist on any particular implementation.)

The idea of changing API was to support other features as well: error codes, logging function, (not in the example above) custom user provided memory allocator.

**Comment by TkTech** [comment]
Apologies, what I get for skimming the thread.

**Comment by lemire** [comment]
> The idea of changing API was to support other features as well: error codes, logging function, (not in the example above) custom user provided memory allocator.

I would argue that we should distinguish two things... runtime dispatch based on CPU features, and API changes.

**Comment by jan-wassenberg** [comment]
Hi Daniel!

> branch on RUNTIME_AVX2 but only, evidently, if the code was compiled with support for AVX2. Again, that's a single predictable branch

Yes, the branch approach works well when there are only 2-3 codepaths. Well-predicted and simple.

> how to make cmake compile multiple versions of the same code with different optimization settings and with different function suffixes

I like the approach taken in PIK (newer link: https://github.com/google/pik/blob/master/pik/simd/targets.h)
```#define TARGET AVX2
#include "impl.cc"
#define TARGET SSE4
#include "impl.cc"
```

impl.cc can use TARGET with token pasting and/or as a value (#if TARGET == SSE4) and/or as a type (Template<TARGET>).

> This requires you to compile the code only once.

Compiling impl.cc multiple times can be useful: it allows reusing the same implementation for SSE4/AVX2 if they only differ by lane count or similar.

> You need to locally tell the compiler not to use fancy instructions, that might be tricky in some cases

Yes, this is a problem. It's important to NOT allow compilers to generate fancy instructions on their own: https://randomascii.wordpress.com/2016/12/05/vc-archavx-option-unsafe-at-any-speed/

**Comment by lemire** [comment]
@jan-wassenberg Thanks for chiming in.

**Comment by lemire** [comment]
We have implemented runtime dispatch in simdjson. 

https://github.com/lemire/simdjson

There is a single compilation. Runtime dispatch works with the single-header version.
